articles/build-rules.md

```diff
@@ -0,0 +1,89 @@
+
+### The Makefile is the build system
+
+Make, rake, Ant, MSBuild, graddle, grunt etc. **are not build systems**<sup>1</sup>. You create build systems with them. 
+
+Jenkins, TeamCity, TravisCI, BuildBot, TFS etc. are build management systems.
+
+## The Rules
+
+### 1. I am the build server
+
+Rule #1 requires that the build server follows the exact same steps as any other developer. 
+
+Expressed the other way around: Every developer has to be able to recreate locally the complete build process without deviations given the development environment and the correct version of the source tree.
+
+### 2. When the build server says no, it means no!
+
+Rule #2 says that if a build server marks a build as broken, then the build is broken. Drop everything and read the logs. 
+
+There is no "it works for me", your build server is Judge Dredd: judge, juror and executioner.
+
+You can only adhere to this rule if you followed Rule #1
+
+### 3. IDEs are the enemy AKA F5 is not a build process
+
+Means that if you drive your development process from an IDE there is no way you can adhere to the Build Rules. 
+
+This rule has major complications on the development environment, ties directly into the subject of allowing your developers to use whatever tools they feel comfortable with and adherence to it is actively sabotaged by Visual Studio.
+
+
+## Adhering to the rules
+
+To create a system that adheres to rules #1 & #2 you have to think like a Lego builder: Lots of small, specialized tasks that do one thing can be used to compose more complex processes etc.
+
+As an example a <em>release</em> task instead of doing everything in one big implementation will depend on the build tasks for each of the libraries and applications and the tasks that run the tests etc. Using the syntax of rake in a contrived example one would do
+
+```ruby
+task :release =>[:"test:all"]
+
+task :"test:all" => [:"test:foo", :"test:bar"]
+
+task :"test:foo" => [:"build:foo"]
+```
+
+Now a developer will probably use the component tasks a lot more than the composite :release task and we will certainly have a build job on the server that only does releases.
+
+This is a necessity since the system needs to satisfy different usage patterns: 
+
+The build server uses composite tasks that implement complete workflows
+
+The developer uses component tasks with surgical precision in the interest of speed and effectiveness.
+
+From the perspective of the build system engineer this approach is self-evident for the same reason it's evident when building applications: Small chunks of code are easier to manage, test, reuse and understand.
+
+All of this segways nicely into the final rule:
```

Comment:
"segways" is not really a verb, isn't it? Is there a more concrete way to put  it`?

Reply:
Sorry, it is a colloquialism. The appropriate grammar here is "segues"